### PR TITLE
Lots of build changes

### LINKED ISSUE #45: Fix Config File Machinery
There's currently some old broken machinery for generating a config file with defines that are set at CMake configure time. That header is no longer included anywhere. It should be fixed and used to provide info about how Galois is configured. Specifically, it should be used to provide:

 - version information
 - the type of aborts used inside the lockable type in the Galois shared object.

Both of those things are currently set as compiler flags in our current CMakeLists, but we shouldn't expect downstream users of an installed library to be aware of the flags used to compile it. That information needs to be included in an installed config header so they can check things with `#ifdef` statements. This is especially important for our operator aborting code because the headers currently will assume exception based aborts by default even though the library builds with setjump/longjump based aborts by default.

### LINKED ISSUE #46: find_package in downstream CMakeLists files should just work
The current install code just installs all the headers and libraries but doesn't create a GaloisConfig.cmake file to make the installed headers immediatly find-able via CMake's `find_package` command. It should do this. There's a minimal working example showing how to do thise at https://gitlab.kitware.com/cmake/community/wikis/doc/tutorials/How-to-create-a-ProjectConfig.cmake-file that we can follow.

### LINKED ISSUE #54: Make option for compiling tools separately
1. Allow tools to be compiled separately
2. When building tools, libllvm should be compiled automatically
3. When building libgalois alone, should not complain about libllvm not found

### PR DESCRIPTION
**Pull request by ddn0**
I can break this up into smaller chunks if people find that easier to review and worth the time.

Closes #54 -- @yishanlu can you confirm that this PR addresses your issue?

Closes #45 -- I opted to not do compiler flags in this PR as it seems (to me) to be pretty hard to support generically. @insertinterestingnamehere if you can point me to what you had in mind for this I can take another look

Closes #46 -- @insertinterestingnamehere I tested this with `add_subdirectory` and with `find_packages` but I may be missing other use cases or edge cases; can you confirm that this is what you wanted?

Open issues:
- Related to #111 but the installation of the libgpu headers is incorrect. libgpu expects a flat namespace for its imports (i.e., with no directory prefixing). I was reluctant to have an explosion of include directories (-Ione -Itwo -Ithree) so I tried to align the libgpu includes with the typical structure that the other libraries have but that means there are subdirectories for each gpu subcomponent (moderngpu, cub). I could have flattened them all, but that seems wrong. Perhaps there is a better way? The previous state was to not install them at all so I consider this PR an improvement.

### CODE REVIEW

**Comment by insertinterestingnamehere** [comment]
Thanks for this! WRT compiler flags, I was referring primarily to the defines. I'll take care of the warning suppression flags myself somewhere else. Things like the architecture we compile for, our build type, and presence of Galois debug info are really just library-specific settings that don't have to be propagated to downstream packages.

**Comment by insertinterestingnamehere** [comment]
The GaloisConfig.cmake.in looks right. I'll test it out shortly. The key thing is just that once Galois is installed, downstream packages should not have to write a custom findGalois CMake setup to find it.

**Comment by yishanlu** [comment]
> Thanks for this! WRT compiler flags, I was referring primarily to the defines. I'll take care of the warning suppression flags myself somewhere else. Things like the architecture we compile for, our build type, and presence of Galois debug info are really just library-specific settings that don't have to be propagated to downstream packages.

With these kinds of flags in a config header and target-based linking, #54 should be resolved.

**Comment by ddn0** [comment]
FWIW, the current set of build errors probably will go away after remove libllvm.

File: lonestar/matrixcompletion/CMakeLists.txt [hunk 1]
```diff
@@ -1,13 +1,18 @@
...
+
+if(CMAKE_COMPILER_IS_GNUCC)
+  target_compile_options(matrixCompletion PRIVATE -ffast-math)
+endif()
 
 find_package(Eigen)
 if(Eigen_FOUND)
-  include_directories(${Eigen_INCLUDE_DIRS})
-  add_definitions(-DHAS_EIGEN -DEIGEN_DONT_PARALLELIZE)
+  target_link_libraries(matrixCompletion Eigen3::Eigen)
```
Comment (insertinterestingnamehere) [inline lonestar/matrixcompletion/CMakeLists.txt:12]: Eigen is header only, so we don't need to worry about linking it.

Comment (insertinterestingnamehere) [reply lonestar/matrixcompletion/CMakeLists.txt:12]: Actually I think the right check for eigen here is:

```
find_package(Eigen3)
if(EIGEN3_FOUND)
  target_compile_definitions(matrixCompletion PRIVATE -DHAS_EIGEN -DEIGEN_DONT_PARALLELIZE)
  include_directories("${EIGEN3_INCLUDE_DIR}")
  MESSAGE("Eigen include dir: ${EIGEN3_INCLUDE_DIR}")
endif()
```

Hopefully the CI systems agree...

**Comment by insertinterestingnamehere** [comment]
In order to be able to find Galois using `find_package`, the generated GaloisConfig.cmake needs to be installed into something like `${CMAKE_INSTALL_PREFIX}/share/Galois/cmake`.

File: libgalois/src/Version.cpp.in [hunk 2]
```diff
@@ -22,14 +22,14 @@
 #define QUOTE(name) #name
 #define STR(macro) QUOTE(macro)
```
Comment (insertinterestingnamehere) [inline libgalois/src/Version.cpp.in:24]: Not necessarily needed in this PR, but we should have a similar set of macros for detecting the version of the Galois headers used at compile time. These functions just tell us the version of the library used.

**Comment by ddn0** [comment]
> In order to be able to find Galois using `find_package`, the generated GaloisConfig.cmake needs to be installed into something like `${CMAKE_INSTALL_PREFIX}/share/Galois/cmake`.

I believe there is some freedom here:

https://cmake.org/cmake/help/latest/command/find_package.html#search-procedure

On my Ubuntu machine, it looks like the vanilla llvm-X-dev package from apt.llvm.org gets packaged under `lib` while system maintainers are expected to put these files into `share`. FWIW, `share/llvm/cmake` is a link to `lib/llvm/cmake`.

```
ddn@spork:usr$ find /usr -name cmake -o -name *Config.cmake
/usr/bin/cmake
/usr/share/cmake-3.13/Modules/FindPkgConfig.cmake
/usr/share/cmake-3.13/Modules/UsePkgConfig.cmake
/usr/share/cmake-3.13/completions/cmake
/usr/share/doc/cmake
/usr/share/cmake
/usr/share/bash-completion/completions/cmake
/usr/lib/llvm-10/cmake
/usr/lib/llvm-10/share/llvm/cmake
/usr/lib/llvm-10/lib/cmake
/usr/lib/llvm-10/lib/cmake/clang/ClangConfig.cmake
/usr/lib/llvm-10/lib/cmake/polly/PollyConfig.cmake
/usr/lib/llvm-10/lib/cmake/llvm/LLVMConfig.cmake
/usr/lib/llvm-10/lib/cmake/llvm/LLVM-Config.cmake
/usr/lib/cmake
/usr/lib/llvm-9/cmake
/usr/lib/llvm-9/share/llvm/cmake
/usr/lib/llvm-9/lib/cmake
/usr/lib/llvm-9/lib/cmake/polly/PollyConfig.cmake
/usr/lib/llvm-9/lib/cmake/llvm/LLVMConfig.cmake
/usr/lib/llvm-9/lib/cmake/llvm/LLVM-Config.cmake
```

File: lonestar/matrixcompletion/CMakeLists.txt (hunk 1 above)
Comment (ddn0) [reply lonestar/matrixcompletion/CMakeLists.txt:12]: I thought one of the benefits of the target-based approach was that header-only libraries can be treated uniformly: https://cmake.org/cmake/help/latest/command/add_library.html#interface-libraries

Though you are right that our internal FindEigen.cmake module doesn't actually export the right symbols here, but I think I'd rather just fix the module than change the method of importing.

**Comment by insertinterestingnamehere** [comment]
Got it. I missed that the config file was getting installed elsewhere. I just refactored my test for using Galois outside of the source tree and it works now.

**Comment by insertinterestingnamehere** [comment]
BTW, this is still marked as WIP, but I think it's good to go. What's left to be done here?

**Comment by ddn0** [comment]
> BTW, this is still marked as WIP, but I think it's good to go. What's left to be done here?

Just rebasing on top of master (since this incorporates the previously unmerged unvendor llvm changes). Let me turn this around shortly.

**Comment by insertinterestingnamehere** [comment]
Sounds great, thanks!

**Pull request merged**